jlm/llvm/backend/rvsdg2dot/rvsdg2dot.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2024 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <jlm/llvm/backend/rvsdg2dot/rvsdg2dot.hpp>
+#include <jlm/llvm/ir/types.hpp>
+#include <jlm/rvsdg/region.hpp>
+#include <jlm/rvsdg/simple-node.hpp>
+#include <jlm/rvsdg/structural-node.hpp>
+#include <jlm/rvsdg/traverser.hpp>
+#include <jlm/rvsdg/type.hpp>
+
+namespace jlm::llvm::rvsdg2dot
+{
+/**
+ * Creates a node in the \p typeGraph representing the given \p type,
+ * or returns such a node if it has already been created.
+ * The function is recursive, and will create nodes for subtypes of aggregate types.
+ */
+static util::Node &
+GetOrCreateTypeGraphNode(const rvsdg::type & type, util::Graph & typeGraph)
+{
+  // If the type already has a corresponding node, return it
+  if (auto * graphElement = typeGraph.GetElementFromProgramObject(type))
+  {
+    auto * node = reinterpret_cast<util::Node *>(graphElement);
+    JLM_ASSERT(node);
+    return *node;
+  }
+
+  auto & node = typeGraph.CreateNode();
+  node.SetProgramObject(type);
+  node.SetLabel(type.debug_string());
+
+  // For all aggregate types, add edges from the nodes representing the children
+  if (auto arrayType = dynamic_cast<const arraytype *>(&type))
+  {
+    auto & elementTypeNode = GetOrCreateTypeGraphNode(arrayType->element_type(), typeGraph);
+    typeGraph.CreateDirectedEdge(elementTypeNode, node);
+  }
+  else if (auto structType = dynamic_cast<const StructType *>(&type))
+  {
+    auto & structDeclaration = structType->GetDeclaration();
+    for (size_t n = 0; n < structDeclaration.NumElements(); n++)
+    {
+      auto & elementTypeNode = GetOrCreateTypeGraphNode(structDeclaration.GetElement(n), typeGraph);
+      typeGraph.CreateDirectedEdge(elementTypeNode, node);
+    }
+  }
+  else if (auto vectorType = dynamic_cast<const vectortype *>(&type))
+  {
+    auto & elementTypeNode = GetOrCreateTypeGraphNode(vectorType->type(), typeGraph);
+    typeGraph.CreateDirectedEdge(elementTypeNode, node);
+  }
+  // TODO: Function types are also aggregate types (arguments and results)
```

Comment:
Could we add function types here as well (and other types we know are missing), and end the cascade with a:

```
else
{
  JLM_UNREACHABLE("Unhandled type");
}
```